tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -5445,97 +5578,84 @@ Status ConvertFullyConnectedHelper(OpConverterParams* params,
   params->converter->SetLayerName(layer, node_def);
   nvinfer1::ITensor* output_tensor = layer->getOutput(0);
 
-  // Reshape output to 1D - this will be a no-op unless using int8 precision.
+  // A fully connected layer produces output with two trailing singleton
+  // dimension. We remove these.
   auto output_dim = output_tensor->getDimensions();
-  output_dim.nbDims = 1;
+  output_dim.nbDims -= 2;
+  // Fill with special value 0, which signals that the corresponding output dims
+  // shall be copied from the input.
+  std::fill(output_dim.d, output_dim.d + output_dim.nbDims, 0);
   TF_RETURN_IF_ERROR(PrepareTensorForShape(
       params->converter, TRT_TensorOrWeights(output_tensor), output_dim,
-      /*validation_only=*/false, &output_tensor, node_def, /*op_instance=*/1));
-
+      /*validation_only=*/false, &output_tensor, node_def,
+      /*op_instance=*/1));
   params->outputs->push_back(TRT_TensorOrWeights(output_tensor));
+  *converted = true;
   return Status::OK();
 }
 
 Status ConvertMatMulHelper(OpConverterParams* params,
                            TRT_TensorOrWeights input_a,
                            TRT_TensorOrWeights input_b, bool transpose_a,
                            bool transpose_b, const NodeDef& node_def) {
-  // TODO: ReorderCKtoKC is currently not general enough to transpose weights
-  // that are not 2D.
-  if ((transpose_a && input_a.is_weights() &&
-       input_a.GetTrtDims().nbDims != 2) ||
-      (transpose_b && input_b.is_weights() &&
-       input_b.GetTrtDims().nbDims != 2)) {
-    return errors::InvalidArgument(
-        "Cannot currently transpose constant input if it is not 2 dimensional");
-  }
-
-  // If A is a tensor, we can only transpose if it is at least 3D in TF,
-  // or TRT will not do the correct transposition.
-  if (transpose_a && input_a.is_tensor() && input_a.GetTrtDims().nbDims < 2) {
-    return errors::InvalidArgument(
-        "Cannot transpose first input if it is a tensor with fewer than 2 "
-        "non-batch dimensions.");
+  if (params->use_implicit_batch) {
+    // In implicit batch mode we are very limited when can we multiply 2D
+    // matrices. If input_A is a 2D tensor, then nbDims==1 (implicit batch dim
+    // not counted). If A is not transposed and B is weight, then we can convert
+    // this treating A as a batch of vectors. This is the only possibility
+    // to implement MatMul with 2D input in implicit batch mode.
```

Comment:
Let's simplify this and not repeat what is obvious from the code:
 In implicit batch mode, we can only implement limited cases of 2D MatMul, through treating A as a batch of vectors.